File: 203.c
#include <stdio.h>
#include <string.h>
#include <stdlib.h>

#include "list.c"

static struct ListNode *rm(struct ListNode *prev, struct ListNode *nd)
{
    // nd!=NULL
    prev->next = nd->next;
    free(nd);
    return prev;
}
struct ListNode *removeElements(struct ListNode *head, int val)
{
    struct ListNode *p;
    struct ListNode *prev = head;
    for (p = head; p; p = p->next)
    {
        if (p->val == val)
        {
            if (prev == p)
            { // head
                head = p->next;
                free(p);
            }
            else
            {
                p = rm(prev, p);
            }
        }
        prev = p;
    }
    return head;
}

int main(int argc, char const *argv[])
{

    int a[] = {1, 2, 6, 3, 4, 5, 6};
    struct ListNode *lst = make_list(a, sizeof(a) / sizeof(a[0]));
    lst = removeElements(lst, 6);
    print_list(lst);
    return 0;
}

File: 203.go
package main

import "fmt"

type ListNode struct {
	Val  int
	Next *ListNode
}

func makeList(a ...int) *ListNode {
	var nd *ListNode
	for i := len(a) - 1; i >= 0; i-- {
		nd = &ListNode{a[i], nd}
	}
	return nd
}
func printList(head *ListNode) {
	for ; head != nil; head = head.Next {
		fmt.Printf("%d->", head.Val)
	}
	fmt.Println()
}
func removeElements(head *ListNode, val int) *ListNode {
	var prev *ListNode
	var p *ListNode
	for p = head; p != nil; p = p.Next {
		if p.Val == val {
			if prev == nil { // head
				head = p.Next
			} else {
				prev.Next = p.Next
			}
		} else {
			prev = p
		}
	}
	return head
}
func main() {
	lst := makeList(1, 1)
	lst = removeElements(lst, 1)
	printList(lst)
}
